### ISSUE TITLE
NBC - dealing with many requests - advice to implementers

### ISSUE THREAD

**Issue by mpiforumbot** [mpi-forum/mpi-forum-historic#86]
**Originally by _htor_ on 2008-12-29 09:49:31 -0600**

---

Chapter 5.13 (draft page 69:44-48):

"Advice to implementors. The use of pipelining may generate many outstanding requests. A high-quality hardware-supported implementation with limited resources should be able to fall back to a software implementation if its resources are exhausted. In this way, the implementation could limit the number of outstanding requests only by the available memory. (End of advice to implementors.)"

**Comment by mpiforumbot** [comment]
**Originally by _bronis_ on 2009-01-02 13:45:10 -0600**

---

I find the second sentence in the current wording confusing. I understand what is meant but only after parsing it multiple times. I think the following rewording would be better:

Advice to implementors. The use of pipelining may generate many outstanding requests. A hardware-supported implementation with limited resources should be able to fall back to a software implementation if its resources are exhausted. In this way, the implementation could limit the number of outstanding requests only by the available memory. (End of advice to implementors.)

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-01-03 10:10:21 -0600**

---

I agree - the wording got a bit confusing with the changes that were proposed during the last meeting :). I think your proposal makes it more understandable. I changed this ticket but not the proposal yet because I expect more discussions on this one as soon as the other reviewers take a look. Please review the new version above!

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2009-01-12 19:02:17 -0600**

---

Don't we need to use the phrase "high-quality" in there somewhere?

  "A **high quality** hardware-supported implementation with limited resources should be able to fall back to a software implementation..."

Otherwise, you'll inevitably create debate of interpretations of this paragraph in the future.  E.g., whether "advice" is mandatory, whether "could" is mandatory, ... indeed, is it mandating that non-blocking collective requests are _only_ limited by available memory?  Aren't there other things that may be desirable to limit NBC requests?  

(yes, I realize I'm being a bit pedantic)

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-01-13 14:05:07 -0600**

---

I added high-quality to the advice. However, limiting the number of outstanding requests too much might have severe portability problems. But I assume that such implementations will fail otherwise :).

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-01-15 14:41:50 -0600**

---

after telecon: added "However, having too many outstanding requests might lower performance." to the companion advice to users. Moved the newly proposed wording to the proposal.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-01-21 08:43:44 -0600**

---

My review:
Current wording is okay with me.

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2009-01-27 11:13:07 -0600**

---

Reviewed. I think it's fine with me.

Jesper

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2009-01-27 19:34:03 -0600**

---

With the companion advice to users phrase (""However, having too many outstanding requests might lower performance."), this is ok with me.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-01-27 19:50:22 -0600**

---

merged into revision 1 - this ticket will be closed/merged into #109.
